source/extensions/transport_sockets/proxy_protocol/proxy_protocol.cc

```diff
@@ -0,0 +1,152 @@
+namespace TransportSockets {
+namespace ProxyProtocol {
+
+ProxyProtocolSocket::ProxyProtocolSocket(Network::TransportSocketPtr transport_socket,
+                                         Network::TransportSocketOptionsSharedPtr options,
+                                         ProxyProtocol_Version version)
+    : transport_socket_(std::move(transport_socket)), options_(options), version_(version) {}
+
+void ProxyProtocolSocket::setTransportSocketCallbacks(
+    Network::TransportSocketCallbacks& callbacks) {
+  transport_socket_->setTransportSocketCallbacks(callbacks);
+  callbacks_ = &callbacks;
+}
+
+std::string ProxyProtocolSocket::protocol() const { return transport_socket_->protocol(); }
+
+absl::string_view ProxyProtocolSocket::failureReason() const {
+  return transport_socket_->failureReason();
+}
+
+bool ProxyProtocolSocket::canFlushClose() { return transport_socket_->canFlushClose(); }
+
+void ProxyProtocolSocket::closeSocket(Network::ConnectionEvent event) {
+  transport_socket_->closeSocket(event);
+}
+
+Network::IoResult ProxyProtocolSocket::doRead(Buffer::Instance& buffer) {
+  return transport_socket_->doRead(buffer);
+}
+
+Network::IoResult ProxyProtocolSocket::doWrite(Buffer::Instance& buffer, bool end_stream) {
+  if (!injected_header_) {
+    auto headerRes = injectHeader();
+    if (headerRes.action_ == Network::PostIoAction::Close) {
+      return headerRes;
+    }
+    injected_header_ = true;
+    auto res = transport_socket_->doWrite(buffer, end_stream);
+    res.bytes_processed_ += headerRes.bytes_processed_;
+    return res;
+  }
+  return transport_socket_->doWrite(buffer, end_stream);
+}
+
+Network::IoResult ProxyProtocolSocket::injectHeader() {
+  if (version_ == ProxyProtocol_Version::ProxyProtocol_Version_V1) {
+    return injectHeaderV1();
+  } else {
+    return injectHeaderV2();
+  }
+}
+
+Network::IoResult ProxyProtocolSocket::injectHeaderV1() {
+  std::ostringstream stream;
+  stream << PROXY_PROTO_V1_SIGNATURE;
+  // Default to local addresses
+  std::string src_address = callbacks_->connection().localAddress()->ip()->addressAsString();
+  std::string dst_address = callbacks_->connection().remoteAddress()->ip()->addressAsString();
+  auto src_port = callbacks_->connection().localAddress()->ip()->port();
+  auto dst_port = callbacks_->connection().remoteAddress()->ip()->port();
+  auto ip_version = callbacks_->connection().localAddress()->ip()->version();
+
+  if (options_ && options_->downstreamAddresses().has_value()) {
+    const auto down_addrs = options_->downstreamAddresses().value();
+    ip_version = down_addrs.version_;
+    src_address = down_addrs.remote_addr_;
+    dst_address = down_addrs.local_addr_;
+    src_port = down_addrs.remote_port_;
+    dst_port = down_addrs.local_port_;
+  }
+
+  Buffer::OwnedImpl buff{};
+  Common::ProxyProtocol::generateV1Header(src_address, dst_address, src_port, dst_port, ip_version, buff);
+  auto res = buff.write(callbacks_->ioHandle());
+  if (!res.ok()) {
```

Comment:
Looking at the raw buffer socket, it also checks if the `res.err` is an `Api::IoError::IoErrorCode::Again`, and returns `KeepOpen` if it is. I wasn't sure if when receiving the `Again` error, it was a possibility that the header was partially sent. Handling that would complicate the logic so for now I've opted to return `Close` on any error. Let me know what you think is the right thing to do here is.

Reply:
@lizan, do you know? Should this do a similar thing to the raw buffer socket, or is it fine to `Close` on any error?